torch/testing/_internal/common_nn.py

```diff
@@ -5033,18 +5028,36 @@ def __init__(self, *args, **kwargs):
         self.with_tf32 = kwargs.get('with_tf32', False)
         self.tf32_precision = kwargs.get('tf32_precision', 0.001)
         self.test_cpu = kwargs.get('test_cpu', True)
+        self.has_sparse_gradients = kwargs.get('has_sparse_gradients', False)
+
+    def _check_gradients(self, test_case, module, input_tuple):
+        params = tuple(x for x in module.parameters())
+        num_inputs = len(input_tuple)
+
+        def fn_to_gradcheck(*inputs_and_params, **kwargs):
+            assert not kwargs
+            return test_case._forward(module, inputs_and_params[:num_inputs])
+
+        # gradcheck doesn't support operators that take in dense inputs but
+        # return sparse parameters. This only happens in the case of nn.Embedding
+        # and nn.EmbeddingBag. Instead, we call `self.check_jacobian`, which
+        # is a slightly different version of gradcheck that can handle this.
+        if self.has_sparse_gradients:
+            assert len(input_tuple) == 1
+            test_input_jacobian = torch.is_floating_point(input_tuple[0])
+            test_case.check_jacobian(module, input_tuple[0], test_input_jacobian)
```

Comment:
nit: the previous version was careful to always pass a tuple as 2nd argument here, but you unpack. Is that the same?

Reply:
Good catch! Yes, it should be the same. The type annotation for check_jacobian accepts `input` as either a Tensor or a Tuple of tensors:
```
check_jacobian(self, module, input: _TensorOrTensors, jacobian_input=True)
```
In fact, `check_jacobian` actually (around 4 months ago) used to only accept input as Tensor, but we added the ability for it to support passing in `input` as a Tensor and as a tuple of Tensors in https://github.com/pytorch/pytorch/pull/44853